### ISSUE TITLE
ets:lookup_element/3 allocation bottleneck if called from multiple processes

### ISSUE THREAD

**Issue by vans163** [erlang/otp#4728]
EDIT: TLDR, :ets.lookup_element for a large term (~60kb) causes memory allocations to spike and the dirty_schedulers to acquire locks trying to reserve ram from the kernel.  The system becomes stalled. This happens when over 7000 processes call into the same table (different rows but similar sized terms).

```elixir
    opt = [
      :public,
      :named_table,
      :ordered_set,
      {:write_concurrency, true},
      {:read_concurrency, true}
    ]
```

**Describe the bug**
We have been noticing large performance regressions when scaling out workloads.  Scaling being non-linear and exponentially poor.  Example 1000 sessions 5% cpu usage, 2000 10%, 5000 50% (critical level), 7000 100% (emulator is over 100%, many processes dont get scheduled).

I have drilled it down to the allocator.  I am not sure why, but allocations are global and it seems 1 (erlang) process is responsible for them, as soon the workload scales beyond the critical level, the global allocator cannot respond in time.  And everything exponentially slows down.

I am not sure on the solution for this, its very hard to find the cause of this as it does not appear on profiling tools when tracing a process.  Also allocator on msacc is pretty low,  the usage gets roilled into emulator/allocator.  

We have EPYC boxes with 64 cores and a high clocked 8core CPU outperforms the EPYC, due to it being able to process the allocations 2x faster which are the bottleneck.

min_heap_size does not seem to be a real solution when the allocation is huge like 2GB+. We have bumped min_heap_size on certain processes and indeed it did help get the allocator portion of msacc down.

**To Reproduce**
A simple way is create a large ets table of like 5000 records of total table size 4gb.  And call :ets.tab2list(),  this will cause a large allocation.   Then reduce the list into a new one.  Probably also would need to simulate a semi-real workload so all the schedulers are busy.

**Expected behavior**
What I would expect to happen is 1 scheduler out of say 64, to be maxed at 100%.

What happens is the entire VM on all 64 schedulers is 100% emulator usage across 64 cores. Starving all the other processes for scheduler time, and the lock seems deep too because the runqueue does not spike too much, but its clear the processes are not getting any scheduler time.

**Affected versions**
Erlang/OTP 23 [erts-11.2] [source] [64-bit] [smp:64:64] [ds:64:64:10] [async-threads:1] [hipe]

Interactive Elixir (1.11.4)

**Additional context**
I could provide some more code / etc but first want to confirm this issue / get some insight on it to craft a appropriate testcase.

**Comment by rickard-green** [comment]
You mention `erts_literal_area_collector` in the title, but what you describe under "To Reproduce" have no direct connection to the literal areas. Literal areas are used for literals in code and terms stored using the [`persistent_term`](https://erlang.org/doc/man/persistent_term.html) API. How come you think literal areas are involved?

When literal areas are removed a global garbage collection of all processes in the system needs to be done. This is scheduled almost simultaneously so you very likely will see a spike in run-queue lengths. Read of data in a literal area is extremely cheap and scales extremely well, at the cost of very expensive removal of literal data. This is intentional. If the literal areas management actually is the cause of your issue, you should reconsider the use of literal data for your use case and store data some other way.

Memory allocations are performed locally on the scheduler that is executing when allocating (at least when it comes to normal schedulers; dirty schedulers is another thing; see later). When deallocating memory it has to be done on the scheduler that allocated it, so the memory block will be sent to the correct scheduler for deallocation. If you deallocate a large number of memory blocks simultaneously, for example by deleting a large ETS table, the individual blocks will be sent for deallocation to the schedulers where they were allocated. That is, multiple schedulers will wake up and perform memory deallocation work if the blocks have been allocated on multiple schedulers. This is intentional in order to get good scalability. Deallocating one large block will never involve more than at most two threads. The thread that wants to deallocate it and the thread that performs the deallocation. ERTS memory allocators used by normal schedulers are in most cases lock free in user space (assuming you have not disabled them). An example of such an allocator is `eheap_alloc` used for allocation of heaps. In order to map memory allocators however use kernel functionality and this functionality may of course use locks though.

Dirty schedulers share a locked allocator instance, so if it is dirty schedulers that are involved they might contend with each other for that lock (normal schedulers never have to acquire this lock though). `msacc` should be able to tell you which types of schedulers are involved. If lots of large processes are garbage collected simultaneously they will be garbage collecting on dirty schedulers and you may get contention on the allocator instance serving dirty schedulers. If you suspect lock contention you can utilize [`lcnt`](https://erlang.org/doc/man/lcnt.html) in order to get more information.

Please provide more information in general about the issue.

Another thing to look out for: do you have transparent huge pages enabled on your system? If so, you want to disable it since it can cause issues like these.

**Comment by vans163** [comment]
I looked at using supercarriers  but could not get it to commit the physical memory on Linux. `+MMsco true +MMscs 67584` it seems to only reserve virtual memory.  I cannot get it to use hugepages either (so 1 large continuous area of about 64G or even 96G is allocated solely to beam). I think I even asked this last year, http://erlang.2086793.n4.nabble.com/How-to-get-HugePages-to-be-used-on-Linux-td4727595.html but could not figure it out.

Ran the system from scratch super happy. With 7k session 16% idle and 28% busywait, so nearly 45% idle, very good.
![image](https://user-images.githubusercontent.com/3028982/114481148-fcff6c00-9bd1-11eb-8804-7cc534a4178e.png)

Now I run the super expensive query (that only runs once per  ~10minutes) that pulls 4gb+ worth of data from ETS tables and reduces it.

1 minute after it fully finishes msacc is like this:
![image](https://user-images.githubusercontent.com/3028982/114481121-ece78c80-9bd1-11eb-8d45-b26ff8ff30f9.png)
![image](https://user-images.githubusercontent.com/3028982/114481086-da6d5300-9bd1-11eb-943d-59f2cb94bbc4.png)

10 minutes after
![image](https://user-images.githubusercontent.com/3028982/114481457-8dd64780-9bd2-11eb-9b21-96c9becd9d2c.png)
![image](https://user-images.githubusercontent.com/3028982/114481670-076e3580-9bd3-11eb-99ad-7b14be420fc6.png)

The system seems even better than before.

For what its worth
![image](https://user-images.githubusercontent.com/3028982/114481772-400e0f00-9bd3-11eb-99fd-9700e5da86c2.png)

I have checked our persistent_term usage, its used solely as a longterm cache in a few small places, all 7k procs have a chance to call it. 
```elixir
{result, expire} = :persistent_term.get(:cached_term, {nil, 0}) 
if time >= expire,
  do: result = expensive_call(); :persistent_term.put(:cached_term, {result, time+minutes10}); result, 
  else: result
```
We create a huge amount of atoms dynamically but they exist in the atom table.

I wonder if there is a better way to benchmark this, I tried eprof on all the session pids (7k), I will try on the entire VM now without the session pids.

I saw perf also produces good results with the BeamAsm jit, but have not tried OTP24 yet.

**Comment by jhogberg** [comment]
> I have checked our persistent_term usage, its used solely as a longterm cache in a few small places, all 7k procs have a chance to call it.
> ```erlang
> {result, expire} = :persistent_term.get(:cached_term, {nil, 0}) 
> if time >= expire,
>  do: result = expensive_call(); :persistent_term.put(:cached_term, {result, time+minutes10}); result, 
>  else: result
> ```

I think this could explain it. There's a gap between checking expiry and updating the cache where any number of processes can sneak in and update the cache too, and this gap will grow wider as the problem worsens.

As `expensive_call/1` pulls in a lot of data there's a good chance that it will have to be GC'd several times on a dirty scheduler before returning. With many processes trying to do the same thing, chances are that the dirty schedulers can't keep up with the workload, further delaying the call to `persistent_term:put/2` that resets the timeout.

Dirty schedulers sharing a single allocator instance (protected with a lock) certainly doesn't help, but from your description I think they would still be up to their neck in work even if the allocation part finished instantaneously.

Try making a single process responsible for updating the cache, which is contacted by the other processes when (and only when) the cache is expired.

As an aside it feels iffy to use `persistent_term` for caches, especially those that expire as quickly as 10 minutes. We envisioned it for things that basically never change after application start and have optimized it accordingly. I hope you know what you're doing.

**Comment by rickard-green** [comment]
I agree with all John said above. Besides that, you want to watch out for processes copying the huge (previous) persistent term when you change it. This by making sure that all processes that have references to the literal (which you are about to change by calling `persistent_term:put()`) drops those references *before* you call `persistent_term:put()`. When garbage collecting literals, any reference to a literal (being removed from a literal area) found will be resolved by copying the literal onto the heap of the process referring to it. In your case, this will cause even more processes with huge heaps, which will be garbage collected on dirty schedulers.

The relatively low cpu usage (~50%) seen while the dirty schedulers are executing full time indicates a resource contention of some kind. This resource is most likely the lock protecting the allocator instance that the dirty schedulers use. You can utilize `lcnt` to check if you want. This lock situation will most likely be improved at some point, but I cannot tell you when this will happen.

At this point everything points to that this issue is caused by how you utilize `persistent_term`.

**Comment by vans163** [comment]
Let me see more today.  So far returned after 8 hours, the node is again in that degraded state. This time 3.0 cpu pressure, so the threads are 300% loaded unable to get work, it seems to be stuck waiting for something as the actual cpu usage is only 33%.  

Alloc on dirties at 99%, alloc on aux at 76%.

If sessions drops down to 4k everything works fine. I will go about removing all the usage of persistent_term.put/delete, so only startup and getters.  But I dont think its this, if it is, there is definitely something not right with deleting a persistent term, it should not cause the entire system to be stuck in what appear to be a alloc deadlock, also its stuck in alloc not GC, i would expect removing a persistent_term to cause a GC.  

The deadlock can be triggered instantly by doing a large ets query + reducing it. Or I guess it happened in the last 8 hours. I wanna get `perf` going to see exactly what the dirties are doing. Will report in a little while.

**Comment by garazdawi** [comment]
Can you annotate (press `a` in perf report, or use `perf annotate`) scheduler_wait to see where exactly in there it is spending all of its time?

**Comment by vans163** [comment]
> Can you annotate (press `a` in perf report, or use `perf annotate`) scheduler_wait to see where exactly in there it is spending all of its time?

2-4k
![image](https://user-images.githubusercontent.com/3028982/114559593-e7be2800-9c39-11eb-8dd9-9a7d539b881e.png)
7k
![image](https://user-images.githubusercontent.com/3028982/114559638-f3a9ea00-9c39-11eb-9059-e852ba190c89.png)

**Comment by garazdawi** [comment]
Do you run the perf reports on a specific thread or all of them? If on all of them, can you show a perf report of only a dirty cpu scheduler? You can use the `t` option in perf report to zoom into just one specific thread.

**Comment by vans163** [comment]
> Do you run the perf reports on a specific thread or all of them? If on all of them, can you show a perf report of only a dirty cpu scheduler? You can use the `t` option in perf report to zoom into just one specific thread.

How should I run the perf analysis `perf report -s pid,sym -i /tmp/perf.data` or `perf report -i /tmp/perf.data`, here is the latter on 1 of the 64 dirties.

libgamemap.so is a rustler pathfinding lib we wrote but again it works fine as long as total sessions is under 5k.

2-4k node
![image](https://user-images.githubusercontent.com/3028982/114575285-03c8c600-9c48-11eb-8c0e-4b74802ce28d.png)

7k node
![image](https://user-images.githubusercontent.com/3028982/114575102-dd0a8f80-9c47-11eb-9241-1ee2b2d91906.png)

**Comment by vans163** [comment]
A little progress report, I removed all the usages of persistent_term.put/2 that could happen across any of the session processes. As well removed some heavy ets usage where not appropriate. It seems the issue has not triggered yet, letting it run.  

EDIT: triggered
Tried running under OTP24 (had to make some small adjustments, :linger socket opt changed to a map from tuple, and crypto apis), with jemalloc on and THP turned off.  It seemed like the system now started allocing physical memory and running 7k sessions after a few hours ran out of ram and the got killed by the distributions oom reaper. So the different I would say here is, the old tests it got into a alloc loop trying to reserve virtual memory constantly, in this latest case using physical memory it gets appropriately killed.

**Comment by vans163** [comment]
Summarized findings so far: using :ets.tab2list when the table is 2GB+ in size (# of objects does not matter), causes systemwide lockups as the allocators try to allocate memory (which is not that much even) and all other work stops. System CPU drops very low waiting for locks.

---

Some observations as I try to get to the bottom of this. I am thinking its a very allocator expensive infinite loop somewhere.  

Some notices, doing a memory expensive :ets.tab2list, over split tables (it was 1 table before), causes the allocator to go haywire in the dirty_schedulers, the entire emulator slows to a crawl while its trying to allocate space.  The system CPU usage drops to near 10%, the busy_wait locks up on all schedulers.

msacc showing a healthy system
![image](https://user-images.githubusercontent.com/3028982/114935334-0e7e8900-9e09-11eb-819a-541996c2860d.png)

:ets.tab2list on 7000 elements sharded across multiple tables of total size 1.9GB (this was 1 table before but as a result of this problem I thought to shard it). Causes this to happen
msacc way stuck allocating
![image](https://user-images.githubusercontent.com/3028982/114936055-f0fdef00-9e09-11eb-97a5-af395f3f9e94.png)
![image](https://user-images.githubusercontent.com/3028982/114936179-112dae00-9e0a-11eb-929c-3b1ef9143bcc.png)

How could reading only 7000 elements, of total size 1.9GB, cause such a systemwide degradation on quite a few dirty schedulers (not just the 1 reading it), my understanding was each scheduler has its own allocator.

**Comment by vans163** [comment]
I found the problem, its basically a :ets.get for a  500kb~ term, called by all 7000 sessions. Causes the dirty_schedulers and aux to lock up in allocate. CPU usage drops considerably as locks are waiting upon to mmap ram from the kernel.

The system memory usage does not increase.

![image](https://user-images.githubusercontent.com/3028982/114942664-e005ab80-9e12-11eb-8c47-d9c5156453f3.png)

**Comment by rickard-green** [comment]
>  I will go about removing all the usage of persistent_term.put/delete, so only startup and getters. But I dont think its this, if it is, there is definitely something not right with deleting a persistent term, it should not cause the entire system to be stuck in what appear to be a alloc deadlock, also its stuck in alloc not GC, i would expect removing a persistent_term to cause a GC.

Well, you're *definitely* wrong! If you abuse the functionality, the runtime system may very well misbehave. It would just be plain wrong to change the functionality to satisfy abuse of it at the cost of use cases that use the functionality as intended.

I'm not sure your use of `persistent_term` is the sole reason for your issues, but it definitely contributes to your problem. Your current solution floods the system with work when repeatedly changing the persistent term.

Also a GC imply memory allocation and deallocation. If all your 7000 processes refer to the old persistent term it will also be 7000 processes garbage collected for each `persistent_term:put()` you do. Your current implementation will also produce multiple `persistent_term:put()` for every `persistent_term:put()` you intended...

> Tried running under OTP24 (had to make some small adjustments, :linger socket opt changed to a map from tuple, and crypto apis), with jemalloc on and THP turned off.

I assume THP is transparent huge pages? In which cases have you used transparent huge pages?

How do you use jemalloc? Both, how did you enable it, and did you disable ERTS allocators?

> How could reading only 7000 elements, of total size 1.9GB, cause such a systemwide degradation on quite a few dirty schedulers (not just the 1 reading it), my understanding was each scheduler has its own allocator.

> I found the problem, its basically a :ets.lookup_element for a 60kb~ term (inside a try/catch, catch never executes the ets key is always there in these cases), called by all 7000 sessions (0.4GB total). Causes the dirty_schedulers and aux to lock up in allocate. 

Each normal scheduler has its own allocator instance that is lock free as long as new raw memory blocks aren't created. A raw memory block is created either from `mmap()`, `malloc()` or ERTS super carrier (if enabled). ERTS super carrier is protected by a lock. How `mmap()` and/or `malloc()` are synchronized depends on the kernel and libc.

All dirty schedulers share a lock protected allocator instance.

Fetching a 60kb term will most likely create a new raw memory block. This will happen on a normal scheduler. Such a large addition to the heap is not unlikely to trigger a GC of the process. If the heap is large enough this will happen on a dirty scheduler which will allocate a new heap on the dirty scheduler. You obviously have a lot of work on dirty schedulers.

Are your system this heavy on dirty schedulers all the time? Memory allocation on dirty schedulers has not been optimized since we did not expect systems to mainly utilize dirty schedulers, but rather utilize dirty schedulers in a few quite rare cases.

A bunch of assorted questions and requests:

* Did your test without `persistent_term` updates work? In general you need to be clearer about what changes you've made with which results.

* Use `system_information:to_file(FileName)` to generate configuration information about the system when you've tested something and report the result here. Attach the produced file.

* Do not use transparent huge pages. At least not while investigating this. We as well as others have on multiple occations seen issues with it when the kernel begin defragmenting pages.

* How much physical RAM and swap do you have on the machine and what are the peeks in usage (physical, virtual, and swap)?

* What OS do you use? Distro, kernel versions etc.

* Describe what your own NIFs do and if they execute dirty or not.

* Use `lcnt` (pointer in my comment above) to see if there are any contention on locks in the runtime system when the system misbehaves.

* Please do not edit comments over and over. It is very hard to follow. Write a new comment unless it is minor spelling mistake or similar.

**Comment by vans163** [comment]
> Also a GC imply memory allocation and deallocation. If all your 7000 processes refer to the old persistent term it will also be 7000 > processes garbage collected for each persistent_term:put() you do. Your current implementation will also produce multiple persistent_term:put() for every persistent_term:put() you intended...

Agreed, at this point I think this was one of the observed problems when I filed with the erts_allocator_collector msgq being so large description. I have since more than a few days ago removed all usages of places where the persistent_term could be set by multiple processes.  This type of bottleneck went away.  (but the 99% alloc on the dirties kept happening so I dug deeper).

> I assume THP is transparent huge pages? In which cases have you used transparent huge pages?  

Yes, I dont think it made much difference. The effect kept happening with them disabled. (default on distro was madvise)

> How do you use jemalloc? Both, how did you enable it, and did you disable ERTS allocators?  

LD_PRELOAD=/jemalloc.so iex ...   
I did not disable ERTS allocators, not sure how.

> Each normal scheduler has its own allocator instance that is lock free as long as new raw memory blocks aren't created. A raw > memory block is created either from mmap(), malloc() or ERTS super carrier (if enabled). ERTS super carrier is protected by a > lock. How mmap() and/or malloc() are synchronized depends on the kernel and libc.

> All dirty schedulers share a lock protected allocator instance.

> Fetching a 60kb term will most likely create a new raw memory block. This will happen on a normal scheduler. Such a large addition to the heap is not unlikely to trigger a GC of the process. If the heap is large enough this will happen on a dirty scheduler which will allocate a new heap on the dirty scheduler. You obviously have a lot of work on dirty schedulers.

I think this is it.  I will try to craft a test case, but the 7000 processes have quite large heaps as they are serializing/deserializing parsing binary messages, many carry total memory of 600kb-1.1MB
![image](https://user-images.githubusercontent.com/3028982/115041638-bb581500-9ea0-11eb-841c-36156c82c207.png)

When this issue was happening, their total memory capped out around 9.9MB, I am guessing each :ets.lookup for a 60KB term hung around somewhere (even tho no references were kept), once it reached 9.9MB, it started trying to get new memory blocks.

```elixir
#causes the allocation lock
map =:ets.lookup_element(Store, uuid, 2)
if time >= (map[:expire_time] || 0) do
  :ets.insert(Store, uuid, Map.put(map, :expire_time, time+60_000*10))
end

#0 issues perfectly smooth version
expire_time = case :ets.select(Store, [{{uuid, %{expire_time: :"$1"}}, [], [:"$1"]}]) do
      [result] -> result
      [] -> 0
 end
if time >=expire_time do
  map = :ets.lookup_element(Store, uuid, 2)
  :ets.insert(Store, uuid, Map.put(map, :expire_time, time+60_000*10))
end
```
Something that comes to mind looking at this is :ets.select_replace being able to do a map put or merge, but thats a sidetrack right now.

> Are your system this heavy on dirty schedulers all the time? Memory allocation on dirty schedulers has not been optimized since we did not expect systems to mainly utilize dirty schedulers, but rather utilize dirty schedulers in a few quite rare cases.

Not heavy at all, around 1% usage on the dirties. (this is the latest version everything is very smooth without that :ets.lookup_element), 7000 sessions.
![image](https://user-images.githubusercontent.com/3028982/115043164-3241dd80-9ea2-11eb-8a21-893e1f1a654a.png)

The system is 70% idle, very happy to see that now. Has been running over 12 hours, rock solid. (43% busywait, 30% idle)
![image](https://user-images.githubusercontent.com/3028982/115043365-703f0180-9ea2-11eb-9db5-fbf44299b072.png)

Based on what you said I think what was happening is these terms were going on the process heap, and not getting cleaned up by GC, eventually reaching 9.9MB then the dirties started allocating. Its possible around 500 processes could be calling the :ets.lookup_element twice per second, copying 60KB each time. 

Maybe one reason this rarely came up is the most common use cases are web servers, the process would read large term from ets then die and the heap would clean up as the GET/POST request dies.  Long running processes for hours might be more rare. But I wonder if this would also happen with something like the new :pg module, since it does a :ets.lookup_element for a list of pids, and that list could be rather large if there is many pids in that group. Maybe list of pids does not cause the same allocations than if say there are binary terms. Our 60kb term is a mix of terms, atoms, integers, floats, some terms of shared bin heap most small bins.

> Did your test without persistent_term updates work? In general you need to be clearer about what changes you've made with which results.

Same problem with or without.

> Use system_information:to_file(FileName) to generate configuration information about the system when you've tested something and report the result here. Attach the produced file.

Let me do that later I need to scrub information out of there, it carries things like the cookie we use for the nodes even (unless I can run it on a empty shell on the same system)

> Do not use transparent huge pages. At least not while investigating this. We as well as others have on multiple occations seen issues with it when the kernel begin defragmenting pages.

Understood. With disabled/madvise set on THP, no difference is observed.

> How much physical RAM and swap do you have on the machine and what are the peeks in usage (physical, virtual, and swap)?

128GB, peak of stable system is 42GB with 7k sessions.  46G virt on BEAM and 38G res on BEAM.  When the issue happens, with jemalloc, res creeps up to 128GB triggering a system reap (systemd oom killer), without it, virt VERY slowly climbs to 128GB, but when the issue is occuring with alloac at 99% in dirties, virt stays put.

0GB Swap

> What OS do you use? Distro, kernel versions etc.

Ubuntu 20.10, kernel 5.8.0

> Describe what your own NIFs do and if they execute dirty or not.

NIF is rustler executes on dirties, a few allocs, they are used for A*.  Running 12 hours (after the fix) no issues with this nif.

> Use lcnt (pointer in my comment above) to see if there are any contention on locks in the runtime system when the system misbehaves.

I used lcnt before this post but did not see any info I understood back when I used it. just showed IIRC pix_lock and maybe alcu_allocator at 150-200% duration. collisions was around 10-30%. If its critical I can run it again.

> Please do not edit comments over and over. It is very hard to follow. Write a new comment unless it is minor spelling mistake or similar.

😅

Let me try to craft a test case now to simulate this.

**Comment by vans163** [comment]
Got it replicated.

```elixir
#iex --sname test

defmodule Alloc do
  def tick() do
    map = :ets.lookup_element(Bench, "1", 2)
    Map.get(map, :expire_time)
    Process.sleep(1000)
    tick()
  end
end

try do
  :ets.new(Bench, [:ordered_set, :named_table, :public, {:write_concurrency,true}, {:read_concurrency, true}])
catch _,_ -> nil end

large_map = Enum.into(1..10000, %{}, fn(idx)->
  {idx, :crypto.strong_rand_bytes(63)}
end)
:ets.insert(Bench, {"1", large_map})

threads = 10000
tasks = Enum.map(1..threads, fn(_)->
    :erlang.spawn(Alloc, :tick, [])
end)


#iex --sname remmer --remsh test@host
:msacc.start(5000)
:msacc.print()
:msacc.stop()
```
![image](https://user-images.githubusercontent.com/3028982/115053471-2c052e80-9ead-11eb-9470-2849067fa4b2.png)

If you comment `Map.get(map, :expire_time)`, a key that does not exist anyways, you get 93% usage in ETS which is correct now.
![image](https://user-images.githubusercontent.com/3028982/115053767-7d152280-9ead-11eb-8876-05b61417940e.png)

**Comment by garazdawi** [comment]
I ran your `ets` benchmark using `lcnt` and there were a lot of conflicts on `alcu_allocator` which would suggest that what Rickard wrote earlier is what is happening.

> If lots of large processes are garbage collected simultaneously they will be garbage collecting on dirty schedulers and you may get contention on the allocator instance serving dirty schedulers. If you suspect lock contention you can utilize lcnt in order to get more information.

The solution to this is to do the allocation of the new heap before re-scheduling. It is something that we plan to do, but have not done yet.

**Comment by rickard-green** [comment]
@vans163 I've created a PR #5187 which makes it possible to enable multiple allocator instances for dirty schedulers. It is based on OTP 24.0.5. You are expected to see an improvement in your case when passing the `+Mdai max` command line argument using this PR.

**Comment by rickard-green** [comment]
Closing this issue. The lock contention issue has been solved by PR #5187 and I have opened an issue #5396 regarding garbage collection improvements that can be made (see also [comment](https://github.com/erlang/otp/pull/5187#issuecomment-970447583) in PR #5187).

**Comment by chasers** [comment]
Thank you for this work. Not entirely sure what I'm doing so terribly wrong but this seems to have fixed it for me.
